var/spack/repos/builtin/packages/phist/package.py

```diff
@@ -27,7 +27,14 @@ class Phist(CMakePackage):
 
     version('develop', branch='devel')
     version('master', branch='master')
-    version('1.9.6', sha256='98ed5ccb22bb98d5b6bf9de0c9960105473e5244978853070b9a3c44138db662')
+    # phist is a required part of spack GitLab CI pipelines.
+    # For them mpich is requested to provide 'mpi'.
+    # phist-1.9.6 changed "use mpi" to "use mpi_f08", which fails with mpich,
+    # which does not appear to provide it when %gcc@7.5.0, resulting in:
+    # Fatal Error: Can't open module file 'mpi_f08.mod' for reading [...]
+    # The safest change it to drop 1.9.6 from the list of versions for now.
+    # To verify future PRs, this command needs to be verified to work for PRs:
+    # spack install phist ^mpich %gcc@7.5.0
     version('1.9.5', sha256='24faa3373003f185c82a658c510e36cba9acc4110eb60cbfded9de370ae9ea32')
```

Comment:
I think this will be easier to model when we have compilers as dependencies (our next big thing to work on after 0.17).

This basically means that 1.9.6 all of a sudden requires Fortran 2008, at least with mpich, which is a large (probably inadvertent) change for a patch version 😬.  That's really a requirement on the compiler used, though in this case it also seems like it's a function of the particular bindings used by mpich: https://lists.mpich.org/pipermail/discuss/2018-March/002203.html.  It looks like gcc 7.5.0 should be able to support it, but doesn't.

I think the right thing to do here is not to delete 1.9.6, but to make 1.9.5 preferred:

```suggestion
    # The safest change (that still allows people to install 1.9.6 if they want) is to
    # prefer 1.9.5 for now. To verify future PRs, this command needs to be verified
    # to work for PRs:
    # spack install phist ^mpich %gcc@7.5.0
    version('1.9.6', sha256='98ed5ccb22bb98d5b6bf9de0c9960105473e5244978853070b9a3c44138db662')
    version(
        '1.9.5',
        sha256='24faa3373003f185c82a658c510e36cba9acc4110eb60cbfded9de370ae9ea32',
        preferred=True,
    )
```

What do you think?

Reply:
Yes definitely prefer the older.

Reply:
@tgamblin @wspear @jthies I've added `preferred=True` for 1.9.5, keeping 1.9.6 available as you requested.

Further fixes:
* As pointed out by @sethrj in https://github.com/spack/spack/issues/26002#issuecomment-932952317 `-fallow-argument-mismatch` is really needed to fix the build with `%gcc@10:` - added:
```py
+        # Workaround for non-compliant and possibly broken code needed for gcc@10:
+        # MPI_Allreduce(localRes, res, 1, MPI_LOGICAL, MPI_LAND, map1%comm, ierr);
+        # MPI_Allreduce(my_ierr,global_ierr,1,MPI_INTEGER,MPI_MAX,map1%comm,mpi_ierr)
+        # Error: Type mismatch between actual argument at (1)
+        # and actual argument at (2) (LOGICAL(4)/INTEGER(4)).
+
+        if spec.satisfies('%gcc@10:'):
+            args.append(define('CMAKE_Fortran_FLAGS', '-fallow-argument-mismatch'))
```

* As @sethrj also pointed out, the git tag 1.9.5 from which we get the tarball has been set to an older commit. It does not have the correct version, so to give the correct version, VERSION_PATCH has to be incremented from 4 (1.9.4) to 5 (1.9.5):
```py
+            # Tag '1.9.5' has moved to an older commit later without fixing the version:
+            filter_file('VERSION_PATCH 4', 'VERSION_PATCH 5', 'CMakeLists.txt')
```
* Parallel build only works since version 1.8.0 (1.7.x failed to compile because of this)
  - added `@propery  def parallel(self): return self.spec.satisfies('@1.8:')` to set it accordingly.

* With `^mpich` all version and gcc compiler combinations I the dual and triple-process (np2 and np3) tests always timed-out, not even 1000 seconds were enough, at least not on a 24-core/48-thread AMD Rome CPU.
  - I added a hint to tell that the tests work with ^openmpi if the tests fail when ^mpi is selected.